pandasaurus_cxg/anndata_analyzer.py

```diff
@@ -0,0 +1,50 @@
+from enum import Enum
+
+import pandas as pd
+
+from pandasaurus_cxg.anndata_loader import AnndataLoader
+from pandasaurus_cxg.schema.schema_loader import read_json_file
+
+
+class AnndataAnalyzer:
+    def __init__(self, file_path: str, schema_path: str):
+        self._anndata_obs = AnndataLoader.load_from_file(file_path).obs
+        self._schema = read_json_file(schema_path)
+
+    def co_annotation_report(self):
+        free_text_cell_type = [key for key, value in self._schema.items() if value]
+        temp_result = []
+        for text in free_text_cell_type:
+            if text in self._anndata_obs.columns:
+                co_oc = (
+                    self._anndata_obs[[text, "cell_type"]].drop_duplicates().reset_index(drop=True)
+                )
+                predicate_dict = co_oc.groupby("cell_type")[text].apply(list).to_dict()
+                co_oc["predicate"] = co_oc.apply(
+                    lambda row: Predicate.CLUSTER_MATCHES.value
+                    if row[text] in predicate_dict.get(row["cell_type"], [])
+                    and len(predicate_dict.get(row["cell_type"], [])) == 1
+                    else (
+                        Predicate.SUBCLUSTER_OF.value
+                        if row[text] in predicate_dict.get(row["cell_type"], [])
```

Comment:
Based on the if/else logic - this looks like adds SUBCLUSTEROF if there are multiple rows with "cell_type" and `text`.  But that could be a subcluster_of relationship in either direction or `overlap`.

Reply:
Let me try to explain my reasoning behind this;
Lets say we have the following structure in predicate_dict:
```
{
'endothelial cell': ['Descending Vasa Recta Endothelial Cell', 'Ascending Vasa Recta Endothelial Cell', 'Afferent / Efferent Arteriole Endothelial Cell', 'Peritubular Capilary Endothelial Cell ', 'Glomerular Capillary Endothelial Cell', 'Degenerative Peritubular Capilary Endothelial Cell', 'Cycling Endothelial Cell', 'Lymphatic Endothelial Cell', 'Degenerative Endothelial Cell'], 
'podocyte': ['Podocyte', 'Degenerative Podocyte'], 
'leukocyte': ['Natural Killer Cell / Natural Killer T Cell', 'M2 Macrophage', 'Neutrophil', 'Monocyte-derived Cell', 'T Cell', 'Plasma Cell', 'Cycling Mononuclear Phagocyte', 'Non-classical Monocyte', 'Classical Dendritic Cell', 'Mast Cell', 'B Cell', 'Plasmacytoid Dendritic Cell', 'Cycling Natural Killer Cell / Natural Killer T Cell']
}
```
I iterate through the df I have and lets say the first `row[text]` is 'Descending Vasa Recta Endothelial Cell' and it corresponds to 'endothelial cell' in the cell type field. I check if 'Descending Vasa Recta Endothelial Cell' is in `['Descending Vasa Recta Endothelial Cell', 'Ascending Vasa Recta Endothelial Cell', 'Afferent / Efferent Arteriole Endothelial Cell', 'Peritubular Capilary Endothelial Cell ', 'Glomerular Capillary Endothelial Cell', 'Degenerative Peritubular Capilary Endothelial Cell', 'Cycling Endothelial Cell', 'Lymphatic Endothelial Cell', 'Degenerative Endothelial Cell']`. Since the length of the list is not 1 I infer 'Descending Vasa Recta Endothelial Cell' as `subcluster_of`  'endothelial cell'

Is there any way to determine the direction of this relationship with the tabular data?

I assumed that everything other than `cluster_matches` and `subcluster_of` should be `cluster_overlaps`, not sure for 100%.